Flow control hazards (#1927)
quicwg/base-drafts: draft-ietf-quic-transport.md

is not excessive buffering at multiple layers.

## Flow Control Deadlocks {#flow-control-deadlocks}

Flow control provides a means of managing access to the limited buffers
endpoints have for incoming data.  This mechanism limits the amount of data that
can be in buffers in endpoints or in transit on the network.  However, there are
several ways in which limits can produce conditions that can cause a connection
to either perform suboptimally or deadlock.

Implementations need to be aware of the possibility of deadlocking and that some
conditions that lead to deadlocking cannot be completely avoided in a protocol
design.

Large messages can produce deadlocking if the recipient does not process the
message incrementally.  If the message is larger than flow control credit
available and the recipient does not release additional flow control credit, a
deadlock can occur.  This is possible even where stream flow control limits are
not reached because connection flow control limits can be consumed by other
streams.

A common implementation technique has flow control credit extended as data is

> read from receive buffers.  In that setting, a length-prefixed message format

Reviewer: I'm confused about why this is so.  The only thing I can think of is if the payload is too big, you can reset/stop_sending the stream.

Author: I'm not following your comment here.

makes it easier for a recipient to leave data unread and thereby withhold flow
control credit.  If flow control limits prevent the remainder of a message from
being sent, a deadlock will result.  However, a length prefix might enable the
detection of this sort of deadlocks.  Where protocols have messages that might

> be processed as a single unit, reserving flow control credit for the entire

Reviewer: Connection level flow control can't be reduced, so are you suggesting it not be further increased until the message arrives?  I haven't implemented this sort of logic, but it seems very tricky to get right and done incorrectly, could result in deadlocks.

Reviewer 2: Concrete example:  You can't consume less than a full QPACK instruction on the encoder stream.  If you read far enough to see that you have a 64KB auth header coming at you, you might logically not continue reading until you've received those 64KB.  If there's not sufficient flow control for the sender to complete sending, nothing gets read and you deadlock.

That might be because the connection flow control never gets increased, or it might be because it got increased but allocated to other streams that have data to send.

Reviewer 3: > Concrete example: You can't consume less than a full QPACK instruction on the encoder stream.

Yes, you can. See [lsqpack_dec_enc_in()](https://github.com/litespeedtech/ls-qpack/blob/5f17c82e8c0f3b2b2faa8984256ff355bc5dc657/src/lsqpack.c#L4186)

Reviewer 4: I think he's just giving an example of an implementation that cannot handle this.  You are correct that the protocol does not mandate this specific behavior.

message atomically ensures that this style of deadlock is less likely.

> Even a message as small as two bytes could produce a deadlock.  Though the

Reviewer: This paragraph isn't adding anything for me.

Reviewer 2: I agree, except that the last sentence is important.

Author: Suggestions sought.  I don't think that I can just leave the last sentence without the preface.

Reviewer 5: How about "A data consumer can read bytes out of a receiver's buffer as they become available to cause the receiver to extend flow control credit to the sender and reduce chances of a deadlock. However, releasing flow control credit might mean that the data consumer might need other means of holding a peer accountable for the state commitment for partially processed messages."

Author: Thanks, that's much better.

chances of deadlock are reduced for smaller messages, the only way to prevent
deadlock is to read every byte as it becomes available.  However, incremental
processing of messages can greatly increase complexity of implementations.
Also, releasing flow control credit might mean that an endpoint might need other
means of holding a peer accountable for the state commitment for partially
processed messages.

Deadlocking can also occur if data on different streams is interdependent.  In
this case, data on one stream arrives before the data on a second stream on
which it depends.  A deadlock can occur if first stream cannot be read and that
prevents flow control limits for the second stream being increased.  To reduce
the likelihood of deadlock for interdependent data, implementations can ensure
that data is not sent until the data it depends on has consumed both stream- and
connection- level flow control credit.

Some deadlocking scenarios might be resolved by cancelling affected streams with

> STOP_SENDING or RST_STREAM, if the protocol permits that.

Reviewer 2: I'd change "if the protocol permits that" to "but depending on the application protocol, this might result in terminating the connection."

is not excessive buffering at multiple layers.

## Flow Control Deadlocks {#flow-control-deadlocks}

Flow control provides a means of managing access to the limited buffers
endpoints have for incoming data.  This mechanism limits the amount of data that
can be in buffers in endpoints or in transit on the network.  However, there are
several ways in which limits can produce conditions that can cause a connection
to either perform suboptimally or deadlock.

Implementations need to be aware of the possibility of deadlocking and that some
conditions that lead to deadlocking cannot be completely avoided in a protocol

> design.

Reviewer 5: I don't think this is accurate. The deadlock problem is not within a protocol here, but in how two protocol implementations (HTTP and QUIC) might interact.

Author: My point is that this is not a matter for protocol design, because no protocol can defend against this effectively in its design and that implementations need to handle the issue.  I'll reword.

... and note that this just adds more weight to moving this to applicability :)

Large messages can produce deadlocking if the recipient does not process the
message incrementally.  If the message is larger than flow control credit

> available and the recipient does not release additional flow control credit, a

Reviewer 5: ```suggestion
available and the recipient does not release additional flow control credit
until the entire message is received and delivered, a
```

deadlock can occur.  This is possible even where stream flow control limits are
not reached because connection flow control limits can be consumed by other
streams.

> A common implementation technique has flow control credit extended as data is

Reviewer 5: ```suggestion
A common flow control implementation technique is for a receiver to extend
credit to the sender as a the data consumer reads data out of the receiver's
buffers. In this setting, a length-prefixed message format
```

> read from receive buffers.  In that setting, a length-prefixed message format

Reviewer 5: ```suggestion
```

> makes it easier for a recipient to leave data unread and thereby withhold flow

Reviewer 5: ```suggestion
makes it easier for the data consumer to leave data unread in the receiver's
buffers and thereby withhold flow
```

control credit.  If flow control limits prevent the remainder of a message from
being sent, a deadlock will result.  However, a length prefix might enable the
detection of this sort of deadlock.  Where protocols have messages that might
be processed as a single unit, reserving flow control credit for the entire
message atomically ensures that this style of deadlock is less likely.

Even a message as small as two bytes could produce a deadlock.  Though the
chances of deadlock are reduced for smaller messages, the only way to prevent
deadlock is to read every byte as it becomes available.  However, incremental
processing of messages can greatly increase complexity of implementations.
Also, releasing flow control credit might mean that an endpoint might need other
means of holding a peer accountable for the state commitment for partially
processed messages.

> Deadlocking can also occur if data on different streams is interdependent.  In

Reviewer 5: ```suggestion
Deadlocking can also occur if a data consumer introduces dependencies between
data sent and received on different streams.
```

Author: I think that I want to keep this abstract, because it's not always the consumer that establishes the dependencies.  QPACK does this intentionally.  It's true that a consumer withholding credit is what causes the deadlock, but I'd rather say that differently.

> this case, data on one stream arrives before the data on a second stream on

Reviewer 5: ```suggestion
Suppose that data on one stream arrives before the data on a second stream on
```

which it depends.  A deadlock can occur if first stream cannot be read and that

> prevents flow control limits for the second stream being increased.  To reduce

Reviewer 5: ```suggestion
prevents the receiver from extending flow control credit for the second stream.
To reduce
```

> the likelihood of deadlock for interdependent data, implementations can ensure

Reviewer 5: ```suggestion
the likelihood of deadlock for interdependent data, the application sender
should ensure
```

> that data is not sent until the data it depends on has consumed both stream- and

Reviewer 5: ```suggestion
that dependent data is not sent until the data it depends on has been accounted
for in both stream- and connection- level flow control credit.
```

> connection- level flow control credit.

Reviewer 5: ```suggestion
```
